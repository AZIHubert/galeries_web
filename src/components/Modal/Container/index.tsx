import * as React from 'react';

import {
  Background,
  Container,
  InnerContainer,
} from './styled';

interface ModalContainerI {
  containerTestId?: string;
  variant?: 'default' | 'danger';
  title?: React.ReactNode;
}

const ModalContainer: React.FC<ModalContainerI> = ({
  children,
  containerTestId,
  variant = 'default',
  title,
}) => (
  <Container
    testId={containerTestId}
  >
    <InnerContainer
      variant={variant}
    >
      {title}
      {children}
    </InnerContainer>
    <Background
      variant={variant}
    />
  </Container>
);

export default ModalContainer;
