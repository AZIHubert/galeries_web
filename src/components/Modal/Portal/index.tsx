import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { CSSTransition } from 'react-transition-group';

import {
  Background,
  Container,
  Fader,
} from './styles';

interface PortalI {
  backgroundTestId?: string;
  callBack?: () => void;
  children: React.ReactNode;
  handleClose: () => void;
  modalTestId?: string;
  open: boolean;
}

const modalRoot = document.getElementById('modal-root');

const Portal = ({
  backgroundTestId,
  callBack,
  children,
  handleClose,
  modalTestId,
  open,
}: PortalI) => {
  const el = React.useRef(document.createElement('div'));

  React.useEffect(() => {
    if (modalRoot && open) modalRoot.appendChild(el.current);
  }, [open]);

  return (
    ReactDOM.createPortal(
      <CSSTransition
        classNames='fade'
        in={open}
        onExited={() => {
          if (modalRoot) {
            modalRoot.removeChild(el.current);
          }
          if (callBack) callBack();
        }}
        timeout={300}
        unmountOnExit
      >
        <Fader>
          <Container
            testId={modalTestId}
          >
            {children}
          </Container>
          <Background
            onClick={handleClose}
            testId={backgroundTestId}
          />
        </Fader>
      </CSSTransition>,
      el.current,
    )
  );
};

export default Portal;
