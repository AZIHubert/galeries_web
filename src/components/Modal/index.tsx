import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { CSSTransition } from 'react-transition-group';

import {
  Background,
  Container,
  Fader,
} from './styles';

interface ModalI {
  callBack?: () => void;
  children: React.ReactNode;
  containerTestId?: string;
  handleClose: () => void;
  modalTestId?: string;
  open: boolean;
}

const modalRoot = document.getElementById('modal-root');

const Modal = ({
  callBack,
  children,
  containerTestId,
  handleClose,
  modalTestId,
  open,
}: ModalI) => {
  const el = React.useRef(document.createElement('div'));

  React.useEffect(() => {
    if (modalRoot && open) modalRoot.appendChild(el.current);
  }, [open]);

  return (
    ReactDOM.createPortal(
      <CSSTransition
        classNames='fade'
        in={open}
        onExited={() => {
          if (modalRoot) {
            modalRoot.removeChild(el.current);
          }
          if (callBack) callBack();
        }}
        timeout={300}
        unmountOnExit
      >
        <Fader>
          <Container
            testId={containerTestId}
          >
            {children}
          </Container>
          <Background
            onClick={handleClose}
            testId={modalTestId}
          />
        </Fader>
      </CSSTransition>,
      el.current,
    )
  );
};

export default Modal;
