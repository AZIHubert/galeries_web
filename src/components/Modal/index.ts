import Container from './Container';
import Portal from './Portal';
import Timer from './Timer';

const Modal = {
  Container,
  Portal,
  Timer,
};

export default Modal;
