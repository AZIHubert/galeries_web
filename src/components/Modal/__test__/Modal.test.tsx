import * as React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {
  fireEvent,
  render,
  screen,
} from '@testing-library/react';

import ThemeProvider from '#contexts/ThemeContext';

import Modal from '../index';

interface ContainerI {
  initialOpen: boolean
}

const innerText = 'Hello world';
const openButtonText = 'open modal';

const Container = ({ initialOpen }: ContainerI) => {
  const [open, setOpen] = React.useState<boolean>(initialOpen);
  const handleClose = () => setOpen(false);
  return (
    <ThemeProvider>
      <button
        onClick={() => setOpen(true)}
      >
        {openButtonText}
      </button>
      <Modal
        open={open}
        handleClose={handleClose}
        testId='modal'
      >
        <div>{innerText}</div>
      </Modal>
    </ThemeProvider>
  );
};

jest.mock('react-transition-group', () => {
  const FakeTransition = jest.fn(({ children }) => children);
  const FakeCSSTransition = jest.fn((props) => (
    props.in ? (
      <FakeTransition>
        {props.children}
      </FakeTransition>
    ) : null));
  return {
    CSSTransition: FakeCSSTransition,
    Transition: FakeTransition,
  };
});

describe('Modal', () => {
  beforeAll(() => {
    // @ts-ignore
    ReactDOM.createPortal = jest.fn((element) => element);
  });
  afterEach(() => {
    // @ts-ignore
    ReactDOM.createPortal.mockClear();
  });
  const handleClose = jest.fn;
  it('renders without crashing', () => {
    const tree = renderer.create(
      <ThemeProvider>
        <Modal open={true} handleClose={handleClose}>
          <div />
        </Modal>
      </ThemeProvider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  describe('when open start at true', () => {
    let child: HTMLElement;
    let modal: HTMLElement;
    let modalBackground: HTMLElement;
    beforeEach(() => {
      const { getByTestId, getByText } = render(<Container initialOpen={true} />);
      child = getByText(innerText);
      modal = getByTestId('modal');
      modalBackground = getByTestId('modalBackground');
    });
    it('should be open', () => {
      expect(child).not.toBeNull();
      expect(modal).not.toBeNull();
      expect(modalBackground).not.toBeNull();
    });
    it('should close when clicking on background', () => {
      fireEvent.click(modalBackground);
      const childAfterClose = screen.queryByTestId(innerText);
      const modalAfterClose = screen.queryByTestId('modalBackground');
      const modalBackgroundAfterClose = screen.queryByTestId('modalBackground');
      expect(childAfterClose).toBeNull();
      expect(modalAfterClose).toBeNull();
      expect(modalBackgroundAfterClose).toBeNull();
    });
  });
  describe('When open start to false', () => {
    let child: HTMLElement | null;
    let modal: HTMLElement | null;
    let modalBackground: HTMLElement | null;
    let openButton: HTMLElement;
    beforeEach(() => {
      const { getByText, queryByTestId, queryByText } = render(<Container initialOpen={false} />);
      child = queryByText(innerText);
      modal = queryByTestId('modal');
      modalBackground = queryByTestId('modalBackground');
      openButton = getByText(openButtonText);
    });
    it('should be close', () => {
      expect(child).toBeNull();
      expect(modal).toBeNull();
      expect(modalBackground).toBeNull();
    });
    it('should open when open is set to true', async () => {
      fireEvent.click(openButton);
      const childAfterOpen = await screen.findByText(innerText);
      const modalAfterOpen = await screen.findByTestId('modal');
      const modalBackgroundAfterOpen = await screen.findByTestId('modal');
      expect(childAfterOpen).not.toBeNull();
      expect(modalAfterOpen).not.toBeNull();
      expect(modalBackgroundAfterOpen).not.toBeNull();
    });
  });
});
