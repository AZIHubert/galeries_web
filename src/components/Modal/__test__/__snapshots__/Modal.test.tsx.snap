// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders without crashing 1`] = `
<div
  className="sc-dlfnbm dqXZeE"
>
  <div
    className="sc-gsTCUz bkImZb modal"
  >
    <div />
  </div>
  <div
    className="sc-bdfBwQ uRgRK background-modal"
    data-testid="modalBackground"
    onClick={[Function]}
  />
</div>
`;
